Separate anchored and unanchored annotations into tabs.

sheetaluk opened the pull request:
Default to the annotations tab when user switches between groups, clears a selection or reloads a page.
The ANNOTATIONS_SYNCED event is emitted when annotations for a document have finished anchoring. Because annotations are not marked as orphans until they fail to anchor, select the annotations tab after they have finished anchoring.

Test cases:
-------------
(1). On page load localhost:3000
-    annotations tab is selected.
-    clicking on new note switches to the notes tab with a new draft created.
-    switching between tabs maintains tab selection, if notes or annotations.
-    when tab selected is orphans, changing groups defaults to the annotations tab.

(2). On page load localhost:3000/#annotations:<id>
-    clicking on new page note should select the notes tab.
-    When an annotation/note is selected, the tab selection carries between group changes.
-    all tests listed in num1 above should apply.

(3). On page load  localhost:3000/#annotations:<orphan_id>
-   clicking on “show all annotations and notes” selects the annotations tab and displays all annotations.
-    all tests listed in num1 and num2 above should apply.
-    switching between groups selects the annotations tab by default


https://trello.com/c/1NVK2jwv/400-add-a-tab-for-unanchored-annotations
https://trello.com/c/OLdLTlLT/342-separate-annotations-and-notes

Hide orphans tab behind orphans_tab flag.
Add tests.

robertknight commented on h/static/scripts/directive/selection-tabs.js:
```diff
@@ -7,21 +7,25 @@ module.exports = function () {
         annotationUI.selectTab(type);
       };
+
+      this.orphansTabFlagEnabled = features.flagEnabled('orphans_tab');
```
There is a potential issue here in that if the value of this feature flag changes, `orphansTabFlagEnabled` will not be updated. The easiest way to resolve this is to make it a function.

robertknight commented on h/static/scripts/root-thread.js:
```diff
@@ -60,15 +60,29 @@ function RootThread($rootScope, annotationUI, features, searchFilter, viewFilter
-      };
+      if (!features.flagEnabled('orphans_tab')) {
+        threadFilterFn = function (thread) {
+          if (state.selectedTab === uiConstants.TAB_ANNOTATIONS || state.selectedTab === uiConstants.TAB_ORPHANS) {
```
If the orphans tab feature flag is not enabled, there should presumably be no way that the orphans tab can be selected? If so, the second condition here is redundant.

robertknight commented on h/static/scripts/test/root-thread-test.js:
```diff
@@ -212,6 +212,19 @@ describe('rootThread', function () {
+      fakeBuildThread.annotation = {$orphan: true};
+
+      fakeAnnotationUI.state = Object.assign({}, fakeAnnotationUI.state,
+        {selectedTab: 'orphan'});
```
This should use the constant for the orphans tab rather than the string literal.

robertknight commented on h/static/scripts/test/root-thread-test.js:
```diff
@@ -212,6 +212,19 @@ describe('rootThread', function () {
       assert.isTrue(threadFilterFn(fakeBuildThread));
     });
 
+    it('generates a thread filter function to match orphans', function () {
```
From the phrasing of these test names, it is a little difficult to understand at first how they relate to what the user sees.

I would suggest something like "filter matches only orphans when Orphans tab is selected" and similar for the other related tests.

robertknight commented on h/static/scripts/test/root-thread-test.js:
```diff
@@ -212,6 +212,19 @@ describe('rootThread', function () {
+      rootThread.thread(fakeAnnotationUI.state);
+      var threadFilterFn = fakeBuildThread.args[0][1].threadFilterFn;
+
+      assert.isTrue(threadFilterFn(fakeBuildThread));
```
I missed this before but this test is broken. The thread filter function takes a _Thread_ as an argument, not the `buildThread` function, which is what _fakeBuildThread_ returns.

This kind of thing is why I would really like to introduce a type checker in our code :)

robertknight commented on h/static/scripts/test/root-thread-test.js:
```diff
@@ -224,6 +237,19 @@ describe('rootThread', function () {
       assert.isFalse(threadFilterFn(fakeBuildThread));
     });
+
+    it('generates a thread filter function for annotations, when all annotations are of type orphans', function () {
```
This test has the same type mismatch as https://github.com/hypothesis/client/pull/52/files#r72415420 and I have the same comment about the naming of the test.

robertknight commented on h/static/scripts/widget-controller.js:
```diff
@@ -97,8 +117,19 @@ module.exports = function WidgetController(
     visibleThreads.setRootThread(thread());
     $scope.selectedTab = annotationUI.getState().selectedTab;
 
+    if (countWaitingToAnchorAnnotations(annotationUI.getState().annotations) > 0) {
```
This can be simplified to `$scope.waitingToAnchorAnnotations = countWaitingToAnchorAnnotations(...) > 0`

robertknight commented on h/static/scripts/widget-controller.js:
```diff
@@ -97,8 +117,19 @@ module.exports = function WidgetController(
+    $scope.totalOrphans = countOrphans(annotationUI.getState().annotations);
+
+    if (!features.flagEnabled('orphans_tab')) {
+      $scope.totalAnnotations = $scope.totalAnnotations + $scope.totalOrphans;
```
It is a bit confusing here that `$scope.totalAnnotations` is assigned twice if the flag is set.

I think this would be clearer as:

```
if (!features.flagEnabled('orphans_tab')) {
  $scope.totalAnnotations = ...
} else {
  $scope.totalAnnotations = ...
}
```

robertknight commented on h/static/scripts/widget-controller.js:
```diff
@@ -312,9 +347,11 @@ module.exports = function WidgetController(
 
-  // When a direct-linked annotation is successfully anchored in the page,
-  // focus and scroll to it
   $rootScope.$on(events.ANNOTATIONS_SYNCED, function (event, tags) {
```
This is a bit confusing, we are handling the `ANNOTATIONS_SYNCED` event in two different places in this code and doing different things in each case - in one place it is aware of the different tabs and in the other place it isn't.

sheetaluk replied:
In widget-controller, we select a tab in 4 cases:
1. page load (select annotations)
2. on group change (select annotations)
3. on group change/page load to a group which has no annotations (select annotations. Also, the ANNOTATIONS_SYNCED event isn't fired in this case).
4. when a user lands on a direct link (select the tab based on the annotation type)

For cases #1 and #4, after the ANNOTATIONS_SYNCED event is fired, we can tell between anchored and unanchored annotations.

robertknight replied:
@sheetaluk - Thanks for the clarification. At the moment I think it is difficult to follow what the business rules are from the code. Having two different listeners set up also means we can run into the problem @finitud is currently looking into where a subscription to some event is set up but then not removed when it should be. When the event is fired in future, a listener that should no longer be triggered ends up handling the event.

The way I would approach this is to first enumerate what the rules are (as you've done above) then try to see if we can simplify them and finally try to express them in code as simply as possible.

Looking at these 4 cases, I think (1-3) can be reduced to one. Whenever we reset the currently loaded set of annotations, the selected tab should be switched to the Annotations tab. I think it should be possible to make that happen without these two event listeners:

1. When the app starts, we can set the initial tab to Annotations (in the initial state in `annotationUI`).
2. When we reset the set of loaded annotations, which happens whenever `loadAnnotations()` is called in `WidgetController` (gah, that's such a stupid name for a class), the tab should be set to `Annotations`.

> For cases 1 and 4, after the ANNOTATIONS_SYNCED event is fired, we can tell between anchored and unanchored annotations. 

In the first case, since we always select the same tab, we don't need to care whether the annotations are orphaned or not do we?

robertknight replied:
To provide a bit more context here, going forwards, I'd like to get to a place where these kinds of rules can be implemented in a more consistent and straightforward way that follows the [basic principles](http://redux.js.org/docs/introduction/ThreePrinciples.html) for a Redux app. Why? Because it makes these kinds of business rules easy to see from the code and test.

This means that all the "when X happens" cases should become actions (calls to `store.dispatch()` in annotationUI. eg. The current group changed, annotations finished loading, annotation anchoring finished) and all the consequences (eg. "tab Y should be selected") are state updates in the reducer functions.

The simpler we can make the logic now, the easier it will be to refactor to this in future.

robertknight commented on h/static/scripts/widget-controller.js:
```diff
@@ -312,9 +347,11 @@ module.exports = function WidgetController(
+    annotationUI.selectTab(uiConstants.TAB_ANNOTATIONS);
+
+    // When a direct-linked annotation is successfully anchored in the page,
+    // focus and scroll to it
```
OK, we now have two cases in which this function could be called that are not accounted for by this comment.

The `ANNOTATIONS_SYNCED` event does not actually mean that the annotation is successfully anchored in the page as this comment says, but rather than attempted anchoring is completed, and we now know whether it was anchored or not.

robertknight commented on h/static/styles/annotation.scss:
```diff
@@ -10,6 +10,10 @@
   background-color: $white;
 }
 
+.annotation-quote-list--orphan {
```
As per the [naming conventions](https://github.com/hypothesis/frontend-toolkit/blob/master/docs/css-style-guide.md#naming) documented in the new CSS style guide, the `--modifier` flags should only be used for static variations of a component. For changing the style of an element depending on some dynamic state, use an `is-$state` class. ie. `.annotation-quite-list.is-orphan`.

The reason for this is that it is obvious which classes are referenced in code and which are not.

robertknight commented on h/static/styles/annotation.scss:
```diff
@@ -10,6 +10,10 @@
 }
 
+.annotation-quote-list--orphan {
+  text-decoration: line-through;
```
Looking at [this screenshot](https://trello-attachments.s3.amazonaws.com/574dae8f7cd863375a5643c8/2180x1100/e62fb94d49354e653faa88e9d289b169/Unanchored_tab.png) from the Trello card, it looks as though the strike-through line for quotes is meant to be a lighter shade of grey than what I see here.

robertknight commented on h/templates/client/viewer.html:
```diff
@@ -1,15 +1,17 @@
-<selection-tabs ng-if="feature('selection_tabs') && !search.query() && selectedAnnotationCount() === 0"
+<selection-tabs ng-if="feature('selection_tabs') && !search.query() && selectedAnnotationCount() === 0 && !waitingToAnchorAnnotations"
```
Reading lines of more than about 100 characters in GitHub's diff viewer is annoying. If a line in a template goes beyond about 80 chars then I would recommend wrapping the attributes onto separate lines as is done below.

robertknight replied:
Hiding the tabs completely until anchoring completes means that there is a flicker where all the tabs disappear when switching groups.

I think that it will look better if the tab bar and Annotations / Page Notes tabs are always shown, but the Orphans tab is only shown if anchoring is complete AND there are orphans.

robertknight commented on h/static/scripts/directive/annotation.js:
```diff
@@ -454,6 +454,10 @@ function AnnotationController(
     return vm.annotation.target;
   };
 
+  vm.orphan = function() {
```
For consistency with the other code here and also for clarity, I would suggest that functions returning booleans should generally have names beginning with `is` or `has`

robertknight commented:
I found two bugs in testing 58eeb00 :

- Clicking the 'New Note' button switches the selected tab to Annotations rather than Notes
- Clicking the 'Reply' button makes all annotations, page notes and orphans disappear from the sidebar

robertknight commented on h/static/scripts/test/root-thread-test.js:
```diff
@@ -199,30 +200,52 @@ describe('rootThread', function () {
+      assert.isTrue(threadFilterFn({annotation: {$orphan: true}}));
+    });
+
+    it('filter matches only notes when Annotations tab is selected', function () {
```
I don't think this test doesn't describes the behavior we want :P Presumably you meant "matches only _annotations_" ;)

robertknight commented on h/static/scripts/test/root-thread-test.js:
```diff
@@ -199,30 +200,52 @@ describe('rootThread', function () {
+      assert.isFalse(threadFilterFn({annotation: {target: [{}]}}));
+    });
+
+    it('filter matches only orphans when Annotations tab is selected', function () {
```
The description here isn't the behavior we want either. We have a test above labeled "filter matches only orphans when Orphans tab is selected" which states the opposite user behavior.

From a user's POV there are two cases:

- When the orphans flag is disabled, the Annotations tab should display annotations and orphans
- When the orphans flag is enabled, the Annotations tab should display only annotations

I think it would be OK to combine the first two cases (orphans flag disabled) into one test with two assertions.

sheetaluk replied:
I changed the descriptions of the tests. I think now they are less confusing :)

sheetaluk replied:
@robertknight I changed the descriptions. I hope its less confusing now.

robertknight replied:
This test description has the same problem as the above test that the description doesn't match the assertion. The test checks that the filter _does not match_ orphans when the Annotations tab is selected.

robertknight commented on h/static/scripts/widget-controller.js:
```diff
@@ -204,9 +234,11 @@ module.exports = function WidgetController(
-        annotationUI.selectTab(
-          tabTypeFromSelection(annotationUI.getState().selectedAnnotationMap, results));
+        // Select appropriate tab - notes, annotations or orphans, for selection
+        $rootScope.$on(events.ANNOTATIONS_SYNCED, function () {
```
Given that we have a `selectTab` call in the _other_ listener for `events.ANNOTATIONS_SYNCED` below, we shouldn't have it here as well.

Something else that you should be aware of is that @finitud recently fixed a set of bugs where listeners were created with `$rootScope.$on` but not unregistered when the controller/directive was destroyed, which happens when the user logs out.

From now on, in controllers and directives, only `$scope.$on` should be used, rather than `$rootScope.$on`. @finitud can provide a more detailed explanation if you'd like.

robertknight commented:
I have pushed a couple of commits that fix the first issue mentioned above in [orphans-fix-reply-handling](https://github.com/hypothesis/client/tree/orphans-fix-reply-handling).

There are couple of issues in the tests that I've noted and an extra `ANNOTATIONS_SYNCED` subscriber which I think needs to be removed.

Otherwise this is looking good and the UX is much better now when switching between groups and waiting for annotations to anchor.

sheetaluk commented:
I fixed the broken 'new page note' button action. I also pulled in @robertknight's changes for fixing the broken reply handling.

>> There are couple of issues in the tests that I've noted and an extra ANNOTATIONS_SYNCED subscriber which I think needs to be removed.
Until annotations are anchored a DLed orphan doesn’t display because we don’t yet know what type the annotation is and we default to selecting the annotation tab.

I have updated the test scenarios in the commit and also below:
Test cases:
-------------
1. On page load localhost:3000
   annotations tab is selected.
   clicking on new note switches to the notes tab with a new draft created.
   switching between tabs maintains tab selection, if notes or annotations.
   when tab selected is orphans, changing groups defaults to the annotations tab.
2. On page load localhost:3000/#annotations:<id>
   clicking on new page note should select the notes tab.
   When an annotation/note is selected, the tab selection carries between group changes.
   all tests listed in #1 above should apply.
3. On page load  localhost:3000/#annotations:<orphan_id>
   clicking on “show all annotations and notes” selects the annotations tab and displays all annotations.
   all tests listed in #1 and #2 above should apply.
   switching between groups selects the annotations tab by default

sheetaluk commented:
Also, I'm not sure what to do about the failing jenkins job :(

robertknight commented:
The second of my commits from the `orphans-fix-reply-handling` branch (341102903e05f2abfe947854caaa59f49e6873eb) which simplifies the tab counts calculations and adds tests for it is missing in this branch.

robertknight commented on h/static/scripts/widget-controller.js:
```diff
@@ -204,9 +242,15 @@ module.exports = function WidgetController(
-        annotationUI.selectTab(
-          tabTypeFromSelection(annotationUI.getState().selectedAnnotationMap, results));
+        // Select appropriate tab - notes, annotations or orphans, for selection
+        $scope.$on(events.ANNOTATIONS_SYNCED, function () {
```
I'm afraid I still don't follow why there are _two_ handlers for `events.ANNOTATIONS_SYNCED` in this module. When this event happens, both handlers will be fired and it isn't obvious when reading the code in which order those events will be fired and consequently which one will "win".

robertknight commented on h/static/scripts/widget-controller.js:
```diff
@@ -167,6 +197,14 @@ module.exports = function WidgetController(
+    if (annotationUI.getState().selectedTab === uiConstants.TAB_ORPHANS) {
+      annotationUI.selectTab(uiConstants.TAB_ANNOTATIONS);
+    } else {
+      annotationUI.selectTab(annotationUI.getState().selectedTab);
```
This line of code doesn't do anything, or at least, I wouldn't expect it to, since all that the `SELECT_TAB` action does is updated the `selectedTab` state variable. If it _is_ doing something, that's a bug and we need to understand what is going on.

robertknight commented on h/static/scripts/widget-controller.js:
```diff
@@ -167,6 +197,14 @@ module.exports = function WidgetController(
     }
   }
 
+  function selectAppropriateTab() {
```
I'm not clear on what 'appropriate' means here.

nickstenning commented on README.md:
```diff
@@ -1,11 +1,3 @@
-[![npm version](https://img.shields.io/npm/v/hypothesis.svg)][npm]
-[![#hypothes.is IRC channel](https://img.shields.io/badge/IRC-%23hypothes.is-blue.svg)][irc]
-[![BSD licensed](https://img.shields.io/badge/license-BSD-blue.svg)][license]
-
```
I'm guessing this wasn't intentional?

robertknight replied:
Ah - I presume this change was made locally while testing out orphan handling by editing the README, which is the test page that the server at http://localhost:3000 serves.

robertknight commented:
I've tested this out locally and taken another look through the code. Aside from a couple of incorrect test descriptions (see comments on outdated diffs), which I've now fixed, this looks good to go.

robertknight commented:
I'll read through this once more with a fresh pair of eyes tomorrow morning and then merge it.

robertknight commented:
I started doing some testing of this feature in the wild this morning on existing pages to see how it looks. I found an issue on https://www.theguardian.com/environment/2016/aug/02/environment-climate-change-records-broken-international-report where it appears that the sidebar stays stuck in the 'waiting for annotations to anchor' state.

Looking at the console errors, my initial suspicion is that anchoring of one or more annotations in the page failed and this never got reported back to the sidebar.

robertknight commented:
This now works as intended, barring the interaction with the existing issue above. I've got a fix in progress for that issue but will submit it as a separate PR.
